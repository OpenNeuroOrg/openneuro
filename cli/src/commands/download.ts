import { Command } from "@cliffy/command"
import type { CommandOptions } from "@cliffy/command"
import { readConfig } from "../config.ts"
import { logger } from "../logger.ts"
import { getRepoAccess } from "./git-credential.ts"

export const download = new Command()
  .name("download")
  .description("Download a dataset from OpenNeuro")
  .arguments("<accession_number> <download_directory>")
  .option(
    "-d, --draft",
    "Download a draft instead of the latest version snapshot.",
  )
  .option(
    "-v, --version",
    "Download a specific version.",
  )
  .action(downloadAction)

export async function downloadAction(
  options: CommandOptions,
  accession_number: string,
  download_directory: string,
) {
  const datasetId = accession_number
  const clientConfig = readConfig()
  const { token, endpoint } = await getRepoAccess(datasetId)

  // Create the git worker
  const worker = new Worker(new URL("../worker/git.ts", import.meta.url).href, {
    type: "module",
  })

  // Configure worker
  worker.postMessage({
    "command": "setup",
    "datasetId": datasetId,
    "repoPath": download_directory,
    "repoEndpoint": `${clientConfig.url}/git/${endpoint}/${datasetId}`,
    "authorization": token,
    "logLevel": logger.levelName,
  })

  console.log("Downloading...")

  worker.postMessage({
    "command": "clone",
  })

  // Close after all tasks are queued
  worker.postMessage({ command: "done" })

  console.log(
    "Download complete. To download all data files, use `datalad get` or `git-annex get`.",
  )
}
