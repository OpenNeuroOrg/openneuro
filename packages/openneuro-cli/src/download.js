import 'cross-fetch/polyfill'
import fs from 'fs'
import path from 'path'
import mkdirp from 'mkdirp'
import { getToken, getUrl } from './config.js'
import { downloadDataset } from './datasets'

export const downloadUrl = (baseUrl, datasetId, tag) =>
  tag
    ? `${baseUrl}crn/datasets/${datasetId}/snapshots/${tag}/download`
    : `${baseUrl}crn/datasets/${datasetId}/download`

export const checkDestination = destination => {
  if (fs.existsSync(destination)) {
    // Exists, check if directory
    if (!fs.lstatSync(destination).isDirectory()) {
      throw new Error(`Target file "${destination}" must be a directory`)
    }
  } else {
    // Doesn't exist, make directory
    fs.mkdirSync(destination)
  }
}

/**
 * Test if a file exists and matches correct size
 * @param {string} destination
 * @param {string} filename
 * @param {number} size
 */
export const testFile = (destination, filename, size) => {
  const fullPath = path.join(destination, filename)
  if (fs.existsSync(fullPath)) {
    return fs.lstatSync(fullPath).size !== size
  } else {
    return true
  }
}

const getFetchHeaders = () => ({
  cookie: `accessToken=${getToken()}`,
})

const handleFetchReject = err => {
  console.error(
    'Error starting download - please check your connection or try again later',
  )
  console.dir(err)
}

/**
 * Download one file from a remote URL
 * @param {string} destination Destination directory path
 * @param {string} filename
 * @param {string} fileUrl URL to download from
 * @param {object} apmTransaction Active APM transaction (optional)
 */
export const downloadFile = async (
  destination,
  filename,
  fileUrl,
  apmTransaction,
) => {
  try {
    const fullPath = path.join(destination, filename)
    // Create any needed parent dirs
    mkdirp.sync(path.dirname(fullPath))
    const writeStream = fs.createWriteStream(fullPath)
    try {
      const response = await fetch(fileUrl, {
        headers: getFetchHeaders(),
      })
      /** @ts-expect-error @type {import('stream').Readable} */
      const stream = response.body
      if (response.status === 200) {
        // Setup end/error handler with Promise interface
        const responsePromise = new Promise((resolve, reject) => {
          stream.on('end', () => resolve())
          stream.on('error', err => {
            if (apmTransaction) apmTransaction.captureError(err)
            reject(err)
          })
        })
        // Start piping data
        stream.pipe(writeStream)
        return responsePromise
      } else {
        console.error(
          `Error ${response.status} fetching "${filename}" - ${response.statusText}`,
        )
      }
    } catch (err) {
      handleFetchReject(err)
    }
  } catch (err) {
    if (apmTransaction) apmTransaction.captureError(err)
    throw err
  }
}

export const getDownload = (
  destination,
  datasetId,
  tag,
  apmTransaction,
  client,
) => {
  const apmSetup = apmTransaction.startSpan('downloadDataset')
  return downloadDataset(client)({ datasetId, tag }).then(async files => {
    apmTransaction.addLabels({ datasetId, tag })
    checkDestination(destination)
    apmSetup.end()
    for (const file of files) {
      if (testFile(destination, file.filename, file.size)) {
        // Now actually download
        const apmDownload = apmTransaction.startSpan(
          `download ${file.filename}:${file.size}`,
        )
        // eslint-disable-next-line no-console
        console.log(`Downloading "${file.filename}" - size ${file.size} bytes`)
        await downloadFile(
          destination,
          file.filename,
          file.urls[file.urls.length - 1],
          apmTransaction,
        )
        if (apmDownload) apmDownload.end()
      } else {
        // eslint-disable-next-line no-console
        console.log(`Skipping present file "${file.filename}"`)
      }
    }
  })
}
