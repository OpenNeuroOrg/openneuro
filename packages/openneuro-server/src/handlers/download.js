import { graphql } from 'graphql'
import schema from '../graphql/schema.js'

const DRAFT_FILES = `
  query dataset($datasetId: ID!) {
    dataset(id: $datasetId) {
      id
      draft {
        id
        files(prefix: null) {
          id
          filename
          size
          urls
        }
      }
    }
  }
`

const SNAPSHOT_FILES = `
  query snapshot($datasetId: ID!, $tag: String!) {
    snapshot(datasetId: $datasetId, tag: $tag) {
      id
      files(prefix: null) {
        id
        filename
        size
        urls
      }
    }
  }
`

export const datasetDownload = (req, res) => {
  const datasetId = req.params.datasetId
  graphql(
    schema,
    DRAFT_FILES,
    null,
    { user: req.user, userInfo: req.userInfo },
    { datasetId },
  )
    .then(({ data }) => {
      res.send({ files: data.dataset.draft.files, datasetId })
    })
    .catch(err => {
      res.status(500)
      res.send(err)
    })
}

export const snapshotDownload = (req, res) => {
  const datasetId = req.params.datasetId
  const tag = req.params.snapshotId
  graphql(
    schema,
    SNAPSHOT_FILES,
    null,
    { user: req.user, userInfo: req.userInfo },
    { datasetId, tag },
  )
    .then(({ data }) => {
      res.send({ files: data.snapshot.files, datasetId, tag })
    })
    .catch(err => {
      res.status(500)
      res.send(err)
    })
}
