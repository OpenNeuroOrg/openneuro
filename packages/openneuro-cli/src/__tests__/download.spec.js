import { downloadUrl, checkDestination } from '../download.js'

jest.mock('../config.js')

let errorSpy
let dirSpy

beforeEach(() => {
  errorSpy = jest.spyOn(console, 'error').mockImplementation()
  dirSpy = jest.spyOn(console, 'dir').mockImplementation()
})
afterEach(() => {
  errorSpy.mockRestore()
  dirSpy.mockRestore()
})

describe('download.js', () => {
  describe('downloadUrl()', () => {
    it('returns snapshot url with tag arg', () => {
      expect(downloadUrl('http://localhost:9876/', 'ds000001', '1.0.0')).toBe(
        'http://localhost:9876/crn/datasets/ds000001/snapshots/1.0.0/download',
      )
    })
    it('returns a draft url with no tag arg', () => {
      expect(downloadUrl('http://localhost:9876/', 'ds000001', undefined)).toBe(
        'http://localhost:9876/crn/datasets/ds000001/download',
      )
    })
  })
  describe('checkDestination()', () => {
    it('throws an error on existing directories', () => {
      expect(() =>
        checkDestination('package.json'),
      ).toThrowErrorMatchingSnapshot()
    })
  })
})
