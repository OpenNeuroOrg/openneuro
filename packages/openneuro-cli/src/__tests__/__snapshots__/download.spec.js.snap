// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`download.js > checkDestination() > throws an error on existing directories 1`] = `[Error: Target file "package.json" must be a directory]`;
